### ISSUE TITLE
Add saints from the Martyrology and their metadata

### ISSUE THREAD

**Issue by tukusejssirs** [pejulian/romcal#208]
Related #154 #20

---

@jarosz mentioned it [once](https://github.com/romcal/romcal/issues/110#issuecomment-564765821) that the saints in the Martyrology for a particular day can be celebrated on that particular day even though it is not in the GRC nor the particular calendar of the place of celebration. Even @emagnier [mentioned](https://github.com/romcal/romcal/pull/169#issuecomment-583752520) in this [PoC - YAML for calendars and Sanctoral experiments (#169)](https://github.com/romcal/romcal/pull/169) that the architecture proposed in that PR ‘could open the possibility to add new features more easily in this area (for example a martyrology).’

IMHO, as the saints listed in the Martyrology could be celebrated around the world on the day they are listed in the Martyrology, when we implement #20 (the _Ordo_, or the Mass celebration rules, as well as the rules for the Liturgy of Hours, for that particular day), it might be a clever idea to output the list of saints from the Martyrology too (of course, as an additional romcal global option that would be set to `false` by default).

Now, I don’t say that this will be implemented anytime soon, but still I think that any development should have this in focus. I wish the the POC mentioned above (#169) would be implemented in the mainline romcal version.

Some resources that would help us accomplish this:
- [_Martyrologium Romanum_, editio typica altera](https://www.vaticanum.com/en/martyrologium-romanum-editio-typica-altera-2004) from 2004, [published in 2005](https://en.wikipedia.org/wiki/Martyrology):
   - this is the latest version of the Martyrology;
   - ~~I’d like to get my hands on this publication, but the price (€75, currently discounted to €67.50) is quite to high for my current financial situation~~ I was gifted with a copy of it in November 2020, so I have already started OCR’ing it [here](https://gitlab.com/tukusejssirs/martyrology) (see the latest [releases](https://gitlab.com/tukusejssirs/martyrology/-/releases));
- [Roman Martyrology 1916](http://www.boston-catholic-journal.com/roman-martrylogy-in-english/Roman%20Martyrology%20by%20Month%20by%20the%20Boston%20Catholic%20Journal.pdf):
   - (??? unofficial) translation of the MR from 1916 edition;
- [The Roman Martyrology 1916](http://www.saintsbooks.net/books/The%20Roman%20Martyrology%20(1914).pdf) ([archive.org](https://web.archive.org/web/20180921074337/http://www.saintsbooks.net/books/The%20Roman%20Martyrology%20(1914).pdf)):
   - a the publication scanned into a PDF;
   - note that although the PDF name states it is from 1914, because it is a translation of the Latin MR1914, but the English MR was published in 1916;
- [eCatholicHub: Catholic Saints Database](https://web.archive.org/web/20120224080055/http://www.ecatholichub.net/study/saints):
   - (??? unmaintained) list of saints from the MR translated into English;
   - the author gave it to [Catholic News Agency](https://www.catholicnewsagency.com/) and they provide access to the [Martyrology](https://www.catholicnewsagency.com/resources/roman-martyrology/martyrology) website, which is searchable, but not quite user-friendly.
   - as the list of the saints/blesseds is not quite easily accessible (and most of the images and biographies are lost), I have created a [Google Sheets table](https://docs.google.com/spreadsheets/d/1huvnEXr9QiCUT3-TuMIxxxG4xggMqG-Vq1E0QwVmBY8/edit#gid=689689355) with all the 6882 saints from [here](https://web.archive.org/web/20120307203958/http://66.37.140.35:3000/ech/View/View_of_saints_print.php?all=1) (the page states that it has 6883, but I have found an empty line in the table between Bl. Casimirus Sykulski and Bl. Liduina (Helisa Angela) Meneguzzi);
- [Martirologio (Italian at vatican.va)](https://www.vatican.va/roman_curia/pontifical_academies/cult-martyrum/martiri/009.html);
- [Martyrolog.ru](http://martyrolog.ru/):
   - an interactive liturgical calendar according the Visigothic/Hispano-Mozarabic manuscripts and printed books of the 6th—16th centuries;
- Martirologio Romano (2001): [Jan-Jun](https://diocesisdecanarias.net/wp-content/uploads/2015/08/santosmartirologioenerojunio.pdf) ([archive.org](https://web.archive.org/web/20200801195244/https://diocesisdecanarias.net/wp-content/uploads/2015/08/santosmartirologioenerojunio.pdf)), [Jul-Dec](http://www.diocesisdecanarias.es/downloads/santosmartirologiojuliodiciembre.pdf) ([archive.org](https://web.archive.org/web/20200801195341/http://www.diocesisdecanarias.es/downloads/santosmartirologiojuliodiciembre.pdf)):
   - Spanish translation of MR2001 (how much it is official and/or accurate, I cannot tell);
- [Martyrologe Romain](https://drive.google.com/file/d/10fSnP5hYbiDbLIZDp-IgLN5G7g0VYeLn/view?usp=drivesdk) (published in March 2020):
   - French translation of MR2004, modified for the needs of the Abbey of Fleury.

**Comment by tukusejssirs** [comment]
A side question: should we directly implement the Martyrology into romcal or would it be better if we create a separate project for the Martyrology (like a library/module) and than source it? If the latter, I would like to know how the data should be formatted. For now (i.e. while the data is gathered), I’ll use a Google Sheets table, but that is rather slow for nearly 7000 different persons (i.e. rows), but on the other hand it is quite good for potential real-time collaboration.

**Comment by CodingAleCR** [comment]
It depends, there are certain festivities that are celebrated globally and others that depend on the region and diocese. For example festivities related to the apostles are usually universal, whereas other saints depend on the region. I think it would be ideal to have the global festivities in romcal, and an additional package for other martyrs, saints and blessed members of our Church. It would ease out maintaining it and adding/changing them wouldn't mean a new release for romcal, making it easier to support them both.

**Comment by tukusejssirs** [comment]
> there are certain festivities that are celebrated globally and others that depend on the region and diocese.

Wnat hou say is true. However, on any day when there is no obligatory celebration (i.e. there is only feria/weekday or optional memorial), one can choose to celebrate any saint (but not blessed) that is listed in the Martyrology on that day.

**Comment by CodingAleCR** [comment]
IMHO, that would be even more reason to keep it separately, that way you would have romcal keeping up with liturgical calendar and on optional memorial or feria/weekday you could query from another library according to your needs or the region you are participating in. You could query that second library for a specific saint, or all saints celebrated at a date or maybe by country/region (this would definitely be a ton of work). Since there are hundreds upon hundreds of blessed, martyrs and saints this would ease up adding/changing the without affecting romcal, ending up in being able to release versions quicker if there's a need for it. I think given the size of the dev team it would be a great approach.

**Comment by tukusejssirs** [comment]
Well, I have no problem with keeping the list of venerated persons (and their metadata) in a separate library, _as long as_ we use only _one singular source_ for them. That said I don’t say that the saints listed on a particular day in the MR2004 would be output by romcal _by default_, but there would be an option (like `outputMartyrologySaints`) which would default to `false` and romcal users could enable it.

As for your implementation proposal, I’d leave it to @emagnier and @pejulian, who as the current coders behind romcal. I like it anyway. :smiley:

---

I think we should discuss the header of the Martyrology DB/table/library/whatever. I’d like it to be as thorough and detailed as possible, while still being simple enough. I always look at it as a DB table, but that might be only my approach.

I think that we should have an information source for each and every (or mostly every) column separately, because each information could be acquired from different source. Though we should try to use official sources, but that is not always possible, therefore there should be three different type of columns for the source information: `source_type`, `source_name` and `source` (while `source` being preferably a URL, or a book citation).

Note that this is just the initial table division draft. We still might divide the `martyrology` table into multiple tables (like a table of sources, where we could list all the sources and link to them from the `martyrology` table) or do any other changes to improve the table structure. However, I think that all the data I list below should be there (one way or another).

Also It might be a good idea to create a list of all dioceses in the world with there name (in Nominative and Genitive), as well as the adjective of their name (in Nominative and Genitive). This list might be either in romcal or in the potential Martyrology library or in a separate library.

---

List of tables:

- `martyrology`:
   - main table for the Martyrology data;
   - it would contain:
      - all the saints and blesseds from the MR2004;
      - all the saints and blesseds that were canonised/beatified since the MR2004 was published;
      - all the saints and blesseds that are venerated in any of the particular calendars (national or religious) in the Roman Catholic Church throughout the world (there are a handful).
   - as Latin is the official language of the RC Church and all names of saints/blesseds are translated _officially_ into Latin, I suggest to change the default language of their names from English to Latin;
   - Our Lady and St Joseph (and possible others) would be listed only once with all the general metadata (for more info, see `multiple_celebrations_of_a_person` table below);
   - for the list of the suggested columns, see below;
- `collective_names`:
   - it would contain a list collective names, which are used for celebration with a list of the saints that are in that groups of saints/blesseds;
   - again, see below for the list of suggested columns;
- `multiple_celebrations_of_a_person`:
   - as Our Lady has many, many celebrations and I’d like to keep the `martyrology` table as clean from duplicates as possible, this table would be used for the all the different celebrations of Our Lady, while pointing to Our Lady from the `martyrology` table;
   - this is also true at least of St Joseph (he has two celebrations: as the Worker and as the Spouse of Our Lady), so he is put here too;
   - for now I won’t create a list of suggested columns for this table;
- `events`:
   - if needed, we could create this table for the list of celebrations which are ‘events’ (i.e. non-person celebrations; e.g. `Shipwreck of St Paul` or `Transfiguration of the Lord` or `Beheading of St John the Baptist`);
   - this is just a thought, but actually it is not part of the Martyrology itself;
   - if we create a separate libary for the Martyrology and if we would _not_ include this in the Martyrology library, this must be kept within romcal;
   - for now I won’t create a list of suggested columns for this table.

---

`martyrology` table columns:

- `id`:
   - a unique ID of the person (we call this `key` in romcal);
- `canonisation_level`:
   - either `saint` or `blessed` (even for multiple persons; multiple person groups should be avoided as much as possible; it would be a great idea to use a separate table for multi-person groups like the `Martyrs of Nowogródek`, where we would join the persons from the main table);
- `first_name`:
   - first name(s) of the person;
- `first_name_gen`:
   - first name(s) of the person in Genitive grammar case;
   - this is not useful for English at all, but for inflected languages (like Latin or Slovak), Genitive is used for the celebration names, as the celebration name is implicitly prefixed by the celebration rank, e.g. `[Solemnity of] Saint Xxx`;
- `last_name`:
   - last name(s) [surname(s)] or patronyms (including `of` if necessary) of the person;
- `last_name_gen`:
   - last name(s) [surname(s)] or patronyms (including `of` if necessary) of the person in Genitive grammar case;
- `name_source_type`:
   - the source type from where we acquired `first_name` and `last_name` (and implicitly the `first_name_gen` and `last_name_gen`);
   - it can be either `official` or `unofficial`;
- `name_source_name`:
   - the name of the source from where we acquired `first_name` and `last_name` (and implicitly the `first_name_gen` and `last_name_gen`);
   - for example, an official source type could be the MR2004, a decree or [AAS](http://www.vatican.va/archive/aas/index_en.htm) (I know of no other official source of information, but that might not be exactly true);
- `name_source`:
   - the source from where we acquired `first_name` and `last_name` (and implicitly the `first_name_gen` and `last_name_gen`);
   - it should be preferably a URL, preferable to vatican.va (or even to vatican.va via archive.org);
   - it could also be a book citation [author[s], book title, publisher, place, year, page(s)], but scanning the relevant pages and uploading there somewhere on the Internet should be always prefered;
- `religious_name`:
   - religious name of the person;
   - note that we usually use the person’s religious name (if they have any) instead of they so called _secular name_ (i.e. `first_names` and `last_names`);
- `religious_name_gen`:
   - religious name of the person of the person in Genitive grammar case;
- `religious_name_source_type`:
   - the source type from where we acquired `religious_name` (and implicitly the `religious_name_gen`);
- `religious_name_source_name`:
   - the name of the source from where we acquired `religious_name` (and implicitly the `religious_name_gen`);
- `religious_name_source`:
   - the source from where we acquired `religious_name` (and implicitly the `religious_name_gen`);
- `native_name_lang`:
   - language of native name of the person;
- `native_name`:
   - native name of the person;
- `native_name_gen`:
   - native name of the person of the person in Genitive grammar case;
- `native_name_source_type`:
   - the source type from where we acquired `native_name` (and implicitly the `native_name_gen`);
- `native_name_source_name`:
   - the name of the source from where we acquired `native_name` (and implicitly the `native_name_gen`);
- `native_name_source`:
   - the source from where we acquired `native_name` (and implicitly the `native_name_gen`);
- `sex`:
   - either `male` or `female` (hopefully, we won’t be required to use `both` and/or `unknown`);
- `titles`:
   - the title(s) of the person used in the `cal1969` liturgy (the New Order of the Mass or the Paul IV Mass);
   - I have no idea if we should keep the titles in one column or 2-3 separate columns, but I’d like to note, that the titles must still be localised by romcal (the titles here should be in Nominative singular, but in some languages, we use Genitive and some saints/blesseds might be celebrated collectively);
- `bishop_of`:
   - in case of the bishops, we should list their (last?) diocese, because in some languages (e.g. German), they always specify the diocese name in the celebration name (`Saints Xxx, bishop of Yyy`);
   - I suggest, that the name of the diocese should be entered in Nominative without the word `diocese` (or even `of`), but that would require us to create a function that would create the na of the diocese; however, if we use the `of Yyy` format, it would look weird from my point of view;
- `other_names`:
   - other names by which the person is celebrated/called (like `St Teresa of Jesus` = `St Teresa of Ávila`);
- `celebration_grc`:
   - the celebration date in `general.ts` if any;
- `celebration_mr2004`:
   - the celebration date in the MR2004 if any;
   - there might come up a problem that some persons might be mentioned on multiple days;
- `celebration_particular`:
   - the celebration dates in the particular calendars if any;
   - I have no idea how to deal with multiple different dates from multiple particular calendars;
   - this column might be dropped, but it is a nice to have feature;
- `celebration_particular_source_type`:
   - the source type from where we acquired `celebration_particular`;
   - I have no idea how to deal with multiple different dates from multiple particular calendars;
   - if `celebration_particular` is dropped, this column should be dropped too;
- `celebration_particular_source_name`:
   - the name of the source from where we acquired `celebration_particular`;
   - I have no idea how to deal with multiple different dates from multiple particular calendars;
   - if `celebration_particular` is dropped, this column should be dropped too;
- `celebration_particular_source`:
   - the source from where we acquired `celebration_particular`;
   - I have no idea how to deal with multiple different dates from multiple particular calendars;
   - if `celebration_particular` is dropped, this column should be dropped too;
- `birth_date`:
   - birth date of the person;
   - keep in mind that the format might be of several types (see [this conversation](https://github.com/romcal/romcal/pull/169#discussion_r376697884); the final formats are still to be defined);
- `birth_place`:
   - where the person was born;
   - it is usually a city, town or village, occasionally a state/country;
- `birth_date_source_type`:
   - the source type from where we acquired `birth_date` and `birth_place`;
- `birth_date_source_name`:
   - the name of the source from where we acquired `birth_date`;
- `birth_date_source`:
   - the source from where we acquired `birth_date`;
- `death_date`:
   - death date of the person;
   - keep in mind that the format might be of several types (see [this conversation](https://github.com/romcal/romcal/pull/169#discussion_r376697884); the final formats are still to be defined);
- `death_place`:
   - where the person died;
   - it is usually a city, town or village, occasionally a state/country;
- `death_date_source_type`:
   - the source type from where we acquired `death_date` and `death_place`;
- `death_date_source_name`:
   - the name of the source from where we acquired `death_date`;
- `death_date_source`:
   - the source from where we acquired `death_date`;
- `mr2004_number_in_day`:
   - at what place the person is mentioned in the MR2004 on the particular day (e.g. in the [MR2001_es](http://www.diocesisdecanarias.es/downloads/santosmartirologiojuliodiciembre.pdf), on 1 July, [St Goulven of Léon](https://en.wikipedia.org/wiki/Goulven_of_L%C3%A9on) is listed on the 6th place);
   - the column name might be name more appropriately, but I couldn’t come up with a better name for now;
- `mr2004_text`:
   - this column might be removed due to copyright reasons;
   - it would contain the text (usually a sentence) from the MR2004 if the person was mentioned there;
   - there might come up a problem that some persons might be mentioned on multiple days;
- `bio_short`:
   - short biography of the person, preferable from the Roman Missal / Liturgy of Hours;
- `bio_short_source_type`:
   - the source type from where we acquired `death_date`;
- `bio_short_source_name`:
   - the name of the source from where we acquired `death_date`;
- `bio_short_source`:
   - the source from where we acquired `death_date`;
- `collective_names`:
   - collective names the person can be / is celebrated in some/all calendars;
- `more_info`:
   - links to websites with more information in different languages;
   - also bibliographical citations of some books;
- `notes`:
   - for any notes, regards, whatever;
- `picture`:
   - a URL to a picture of the person.

`collective_names` table columns:
- `id`:
   - a unique ID of the collective_name (we call this `key` in romcal);
- `name`:
   - the collective name;
- `number`:
   - how many persons are included in the collective name
- `definitie_article`:
   - if the collective name requires a definite article in English;
   - this column might be dropped, as usually all collective names use definite article in English, but then this must be took into account in romcal;
- `notes`:
   - for any notes, regards, whatever.

**Comment by tukusejssirs** [comment]
Another thought: as we want to have localised persons (and event celebrations) and their metadata, we might need either (1) all the suggest tables localised in all (supported) languages; or (2) we should have one master set of those tables (e.g. in Latin) and then we should have just specific data localised (like `en_names` table for all names localised in English, `en_places` for all places localised in English, etc). I think that the second option is better due to eliminating duplicate data and thus minimising the size.

**Comment by emagnier** [comment]
Martyrology is based on a civil calendar, not on a liturgical calendar. So I agree with @CodingAleCR to keep it separately from Romcal. However it could eventually share some of the localisation resources files with Romcal someday, when we will have time to improve this part with an i18n library.

**Comment by tukusejssirs** [comment]
> Martyrology is based on a civil calendar, not on a liturgical calendar.

I don’t fully understand this comment of yours—what does it have to do with this issue?

From my point of view, generally, the MR2004 is a list of saints/blesseds on each day of a year (a civil year if you wish) with a sentence or two about their lives. Now, what I want to do is a DB table(s) that would contain (in the most basic form) the names of the saints/blesseds with their dates they are ascribed to and this would be used in romcal instead of the `src/locales/*.ts` files. For joint celebrations, we would create a separate table that would reference the persons from the main table.

> So I agree with @CodingAleCR to keep it separately from Romcal.

I have no problem with that, as I already said.

> However it could eventually share some of the localisation resources files with Romcal someday […]

Now, this is the part which talks directly the reason why this particular issue (#208) was opened in the first place: I want to know how to create the list of saints (based on the MR2004) that would make it easier to be used by romcal. I know that this is just a preparation, because the main work (creating the list of the saints) will take some time, but I’d like to hear from you how would it suite you/romcal to be structured and/or implemented.

The reason why I want to create the list is not the Martyrology itself, but romcal. I still wish the #154 issue would be fixed _as soon as possible_, because (while creating/localising `la.ts`) I have found another 90+ individual issues (even duplicates) I still need to tend to. These issues are a never-ending story. The proper fix for these issues (and a way to prevent them in future) is to create a full list of saints/blesseds who _might_ be venerated in any of the particular calendars.

> […] when we will have time to improve this part with an i18n library.

I wish it would be more important to you.

**Comment by tukusejssirs** [comment]
I have just finished my upgraded proposal on the structure of tables (and the relations between them).

First of all, I moved all eCatholicHub sheets (as well as some companions list sheets) from the [original Google Sheets](https://docs.google.com/spreadsheets/d/1huvnEXr9QiCUT3-TuMIxxxG4xggMqG-Vq1E0QwVmBY8/edit?usp=sharing) to a [separate one](https://docs.google.com/spreadsheets/d/1cjF0LhNR4yR901q58naAmO63x99n2Vh6feei9-_CzmE/edit?usp=sharing). From now on, the [former table](https://docs.google.com/spreadsheets/d/1huvnEXr9QiCUT3-TuMIxxxG4xggMqG-Vq1E0QwVmBY8/edit?usp=sharing) contains romcal sheets only and this is the table I’ll talk about in the rest of this comment.

---

The [Google Sheet](https://docs.google.com/spreadsheets/d/1huvnEXr9QiCUT3-TuMIxxxG4xggMqG-Vq1E0QwVmBY8/edit?usp=sharing) contains the following sheets:
- [`notes`](https://docs.google.com/spreadsheets/d/1huvnEXr9QiCUT3-TuMIxxxG4xggMqG-Vq1E0QwVmBY8/edit#gid=1599512798):
   - Here I listed some remarks to the tables.
- [`tables`](https://docs.google.com/spreadsheets/d/1huvnEXr9QiCUT3-TuMIxxxG4xggMqG-Vq1E0QwVmBY8/edit#gid=1133780585):
   - Here I listed all the tables I suggest to create.
   - Basically, there is one main table (`martyrology`), which should contain all that should not be localised and references to other (auxiliary) tables.
- [`potentially_missing_columns`](https://docs.google.com/spreadsheets/d/1huvnEXr9QiCUT3-TuMIxxxG4xggMqG-Vq1E0QwVmBY8/edit#gid=251311597):
   - A list of columns, of which I’m either not sure _if_ or _how_ they should be implemented.
- and a sheet per table (from `tables`) with the table header (or columns if you wish) definition.

---

Any suggestions? :smiley:

**Comment by tukusejssirs** [comment]
I just want to let you know I was gifted with [MR2004](https://www.vaticanum.com/en/martyrologium-romanum-editio-typica-altera-2004). 🎉 🙏 I am about to scan it and OCR it in order to ease the processing, ultimately filling in the data in the [Google Sheets tables](https://docs.google.com/spreadsheets/d/1huvnEXr9QiCUT3-TuMIxxxG4xggMqG-Vq1E0QwVmBY8/edit?usp=sharing). Of course, it’ll take some time. :wink:

**Comment by tukusejssirs** [comment]
> @emagnier [#247 (comment)](https://github.com/romcal/romcal/issues/247#issuecomment-843516114)
> 
> 1. Within the romcal repo, data will always be stored in files (`.ts`, or even better in `.json` or `.yaml` files), to keep simplicity and make contribution easy. And indeed we will have to find an easy way to reduce file size for environments that require it (front-end/browser).

Well, my idea is to move all l10n data to a separate repo (`sanctoral`). Even you talk about modularisation in the 2nd point (and I believe we talked about this some time ago) below and this is my proposal. I this that even celebration data (`src/calendar/*.ts`) should be moved to separate repo, and also the calendar version/edition (`cal1962`, `cal1969`; a repo per calendar), but from my point of view that is a separate issue and should ve solved later.

That is why I think we should concentrate on choosing best possible way to put the l10n data (the tables) in an module/package external to romcal (it might and probably should be in the romcal org though) However, ATM I am not sure what is the best way to such data, which will be huge. Therefore we need to find away to make it possible for romcal to extract the specific data at build time required by the builder.

> 2. Be able to connect a database like MongoDB should be possible if we redesign the l10n management (a more modular architecture) and ease the import/export data from APIs. But this might not be the first priority.

That is thought about. The thing is, is MongoDB the best way to store such data? Will it be easily localisable (via a tool like POedit, Transifex, Crowdin, custom tool)?

> 3. If you already use Google Spreadsheet, I guess you can continue to work with it for now (don't forget to do offline backup time to time). Then data should be easy to extract from a well organized spreadsheet.

The thing is, I’d like to do it onceband properly, i.e. in the tool we (right now) think is the best for our use case.

> 4. Last point, but the most important: a big thank you for all of this hard work you are doing! This is very valuable!

Thanks, @emagnier! ❤️ This is a love project for me. I still wish I could help more on the coding side (maybe one day). :wink:

**Comment by tukusejssirs** [comment]
Just to let everyone know of the progress of OCR’ing MR2004. [Here](https://gitlab.com/tukusejssirs/martyrology) is the repository. You also might want to look at the [releases](https://gitlab.com/tukusejssirs/martyrology/-/releases) and [`changelog.md`](https://gitlab.com/tukusejssirs/martyrology/-/blob/master/changelog.md).

I’m sure it is far from perfect. Anyone who is willing to help me on any front is welcome. :wink:

Of course, after I proofread the whole book, I’ll start to fill up the [tables](https://docs.google.com/spreadsheets/d/1huvnEXr9QiCUT3-TuMIxxxG4xggMqG-Vq1E0QwVmBY8).
